### PR TITLE
[CA-1687] Upgrade Client Java by upgrading workbench google and google2

### PR DESCRIPTION
**Pull request by s-rubenstein**
Ticket: [Link to Jira ticket](https://broadworkbench.atlassian.net/browse/CA-1687)
This upgrades our workbench libs dependencies. There are a few changes to the tests that need to be made in order to deal with increased inefficiencies or removed test functionality.

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://broadinstitute.github.io/dsp-appsec-security-risk-assessment/) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala [hunk 1]
```diff
@@ -38,11 +38,36 @@ import org.scalatest.matchers.should.Matchers
 import org.scalatest.{BeforeAndAfterAll, PrivateMethodTester}
 import org.scalatestplus.mockito.MockitoSugar
 
-import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.ExecutionContext.Implicits.{global => globalEc}
 import scala.concurrent.Future
 import scala.concurrent.duration._
 import scala.util.{Success, Try}
 
+class SamMockGoogleDirectoryDAO extends MockGoogleDirectoryDAO {
```
Comment (s-rubenstein) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala:47]: This exists because at some point the MockGoogleDirectoryDAO lost the lower method. The strictly correct fix would be to go into workbench libs and make the change there. I want to follow up with that and then go back and change this, because I want to move as fast on the java upgrade as possible, since its chance to go stale is very high.

File: build.sbt [hunk 2]
```diff
@@ -21,4 +21,13 @@ sys.env.getOrElse("JAVA_OPTS", "").split(" ").toSeq.map { opt =>
   javaOptions in reStart += opt
 }
 
+inThisBuild(
```
Comment (s-rubenstein) [inline build.sbt]: Used to run the scalafix. I can remove if we don't want it still in there.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala [hunk 3]
```diff
@@ -35,7 +35,7 @@ class GoogleGroupSyncMonitorSpec(_system: ActorSystem) extends TestKit(_system)
   }
 
   "GoogleGroupSyncMonitor" should "handle sync message" in {
-    implicit val patienceConfig = PatienceConfig(timeout = 1 second)
+    implicit val patienceConfig = PatienceConfig(timeout = 1 minute)
```
Comment (andy7i) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala]: why this change?

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala (hunk 1 above)
Comment (andy7i) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala:47]: are we tracking this anywhere?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala [hunk 4]
```diff
@@ -160,8 +159,8 @@ object Boot extends IOApp with LazyLogging {
           // Use resourceNamePrefix to avoid collision between different fiab environments (we share same firestore for fiabs)
           val lock =
             DistributedLock[IO](s"sam-${config.googleServicesConfig.resourceNamePrefix.getOrElse("local")}", appConfig.distributedLockConfig, ioFireStore)
-          val newGoogleStorage = GoogleStorageInterpreter[IO](googleStorage, Blocker.liftExecutionContext(blockingEc), None)
-          val googleKmsInterpreter = GoogleKmsInterpreter[IO](googleKmsClient, blockingEc)
+          val newGoogleStorage = GoogleStorageInterpreter[IO](googleStorage, blockerBound = None)
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:162]: is this going from blocking to non-blocking (`blockerBound = None`)?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala [hunk 5]
```diff
@@ -124,6 +126,7 @@ class GoogleGroupSyncMonitorActor(
 
   override def receive = {
     case StartMonitorPass =>
+      logger.debug(s"Pulling messages off: $pubSubSubscriptionName")
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala:129]: what is this change / how does this change relate to the ticket?

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/TestSupport.scala [hunk 6]
```diff
@@ -49,8 +51,7 @@ trait TestSupport{
   def runAndWait[T](f: IO[T]): T = f.unsafeRunSync()
 
   implicit val futureTimeout = Timeout(Span(10, Seconds))
-  implicit val cs = IO.contextShift(scala.concurrent.ExecutionContext.global)
-  implicit val timer = IO.timer(scala.concurrent.ExecutionContext.global)
+//  implicit val timer = Temporal[IO]
```
Comment (andy7i) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/TestSupport.scala]: remove?
```suggestion

```

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala [hunk 7]
```diff
@@ -38,11 +38,36 @@ import org.scalatest.matchers.should.Matchers
...
+  override def lockedDownGroupSettings = new Groups()
+    .setWhoCanAdd("ALL_OWNERS_CAN_ADD")
+    .setWhoCanJoin("INVITED_CAN_JOIN")
+    .setWhoCanViewMembership("ALL_MANAGERS_CAN_VIEW")
+    .setWhoCanViewGroup("ALL_OWNERS_CAN_VIEW")
+    .setWhoCanInvite("NONE_CAN_INVITE")
+    .setArchiveOnly(
+      "true"
+    ) // .setWhoCanPostMessage("NONE_CAN_POST") setting archive only is the way to set it so no one can post
```
Comment (andy7i) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala:56]: should `.setWhoCanPostMessage("NONE_CAN_POST")` be in this comment, or should it be on the next line like:
```suggestion
    ) // setting archive only is the way to set it so no one can post
    .setWhoCanPostMessage("NONE_CAN_POST")
```

File: project/build.properties [hunk 8]
```diff
@@ -1 +1 @@
-sbt.version=1.4.7
+sbt.version=1.5.0
```
Comment (andy7i) [inline project/build.properties]: why 1.5.0? / did you try the latest (I think 1.6.2)?

File: build.sbt (hunk 2 above)
Comment (andy7i) [reply build.sbt]: I think we should remove it once its usefulness has expired. Is it useful beyond the cats migration?

**Review by andy7i — COMMENTED** [review]
overall, looks good. Thanks for taking on this monster!

(minor) - might be helpful to have signposts for areas we should review, especially when the PR has 48 files

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala (hunk 4 above)
Comment (gpolumbo-broad) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:162]: I'm not sure what the right answer is here, but this might be helpful in figuring it out: https://typelevel.org/cats-effect/docs/migration-guide#blocker

**Review by gpolumbo-broad — APPROVED** [review]
All comments/questions I have, @andy7i already asked.  So long as those get addressed: 👍

**Comment by gpolumbo-broad** [comment]
Leaving this link here since I didn't see it elsewhere:  https://typelevel.org/cats-effect/docs/migration-guide might be helpful for this PR and possibly if we look back at this in the future if any bugs crop up as a result.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala (hunk 3 above)
Comment (s-rubenstein) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala]: We think there's some inefficiencies that were introduced, and it is hard to tell what is going to be happening on a machine. On my machine it took between .75 and 2 seconds to run, so Edmund and I thought it was worth just bumping to a minute to give it sufficient time to run.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala (hunk 4 above)
Comment (s-rubenstein) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:162]: No, its just naming the None which was the last arg on line 163. No change to the behavior, just naming the param

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala (hunk 5 above)
Comment (s-rubenstein) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala:129]: We added this to test the fix, and since its a debug I think it would be worth leaving there.

File: project/build.properties (hunk 8 above)
Comment (s-rubenstein) [reply project/build.properties]: I updated to the minimum possible. Because its SBT I would prefer to keep version changes as small as possible in order to reduce risk of changes to our commands - especially while we are already doing a big upgrade.

My general rule is one big upgrade at a time. If we want to upgrade this another major version, we can, but I wouldn't want to do that at the same time as this upgrade in order to keep things as small as possible

**Comment by s-rubenstein** [comment]
@gpolumbo-broad that doc is what I actually followed to do this migration

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala (hunk 1 above)
Comment (s-rubenstein) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala:47]: https://broadworkbench.atlassian.net/browse/CA-1803

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala (hunk 7 above)
Comment (s-rubenstein) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala:56]: It should be in this comment - this is copied directly from the GoogleDirectoryDAO implementation file, without any changes.

**Comment by s-rubenstein** [comment]
In general @andy7i I would agree with you on pointing out places to review, but when it comes to a java upgrade (and major library like Cats) I think it is worth taking a look at every little change because that's kind of the point of the upgrade.

If I had to pick places to take extra eyes, the places where I left individual comments are where I would take a look first. Those are where I think there's more trickiness/there was a decision point behind it.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala (hunk 4 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:162]: looks like we don't need to name it if we don't want to. No strong preference there

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala (hunk 7 above)
Comment (andy7i) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala:56]: weird!

File: project/build.properties (hunk 8 above)
Comment (andy7i) [reply project/build.properties]: feels like we really **really** need something like scala steward

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala (hunk 3 above)
Comment (andy7i) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala]: did you consider the possibility that we might miss significant performance issues due to this change?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala (hunk 4 above)
Comment (s-rubenstein) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:162]: We definitely don't need to, but both Edmund and I agreed that naming it was more clear - it makes it easier to see what the parameter is at a glance.

**Comment by andy7i** [comment]
> In general @andy7i I would agree with you on pointing out places to review, but when it comes to a java upgrade (and major library like Cats) I think it is worth taking a look at every little change because that's kind of the point of the upgrade.

Were any of the changes made by scalafix? I think we could ignore most changes made by scalafix.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala (hunk 3 above)
Comment (s-rubenstein) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala]: Not sure - I don't think its significant because we didn't see other test failures, and the scale was low, but it is something we should keep an eye on for perf tests.

**Comment by s-rubenstein** [comment]
The only changes made by scalafix were the Timer->Temporal and removing the contextshift stuff. The imports were done by hand, which is the majority of the file changes, and should probably be looked at, at least briefly.

File: project/build.properties (hunk 8 above)
Comment (s-rubenstein) [reply project/build.properties]: I agree

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala (hunk 3 above)
Comment (andy7i) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala]: I like the idea of making a change here so that we don't have to remember this / keep an eye on it. So I think I'd prefer a shorter timeout (maybe 5-10 seconds) rather than 1 minute.

**Review by andy7i — APPROVED** [review]
1 last comment but not blocking if you think it's not necessary.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala (hunk 3 above)
Comment (s-rubenstein) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitorSpec.scala]: Makes sense. I'll drop it to 5 seconds, that should be fine

File: project/build.properties (hunk 8 above)
Comment (s-rubenstein) [reply project/build.properties]: Did you know `sbt build` isn't a command? I didn't, for some reason I thought that was the command to use - either way, the one sbt command that we use in Sam that is in our readme is `sbt test`, which worked.
